crates/biome_string_case/src/lib.rs

```diff
@@ -447,48 +447,81 @@ pub trait Collator {
         let mut iter2 = iter2.into_iter();
         loop {
             match (iter1.next(), iter2.next()) {
-                (Some(c1), Some(c2)) if c1 == c2 => {}
+                (None, None) => {
+                    // All characters of `iter1` and `iter2` are equal.
+                    return Ordering::Equal;
+                }
+                (None, Some(_)) => {
+                    // `iter1` is a prefix of `iter2`.
+                    return Ordering::Less;
+                }
+                (Some(_), None) => {
+                    // `iter2` is a prefix of `iter1`.
+                    return Ordering::Greater;
+                }
+                (Some(c1), Some(c2)) if c1 == c2 => {
+                    // For now, all characters of `iter1` and `iter2` are equal.
+                }
                 (Some(mut c1), Some(mut c2)) => {
-                    if let (Some(n1), Some(n2)) =
-                        (self.as_ascii_digit(&c1), self.as_ascii_digit(&c2))
```

Comment:
The main part of the bug was coming from here.
This is too aggressive as condition. I mainly added it as a way of avoiding the loop.
Now we always enter in the loop.